###  Contents
###### 1.0 Importing Libraries
###### 2.0 Importing product dataset
###### 3.0 Checking the columns with mix datatype
###### 4.0 checking for the column that contains null
###### 5.0 Creating subset of the dataframes 
###### 6.0 Checking for duplicate 
###### 7.0 Importing orders dataset
###### 8.0 Checking for the missing number in dataframe
###### 9.0 using fillna for inputation 

### 1.0 Importing Libraries

In [4]:
import pandas as pd
import numpy as np
import os

### 2.0 Importing product dataset                     

In [10]:
df_prods = pd.read_csv('/Users/shaqurcasco/Downloads/ Instacart Basket June: 26: 2023 CF/3 SCRIPT/CF Achievement 4/CF_products.csv')


In [11]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [12]:
#Checking the information of the product dataframe
df_prods.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49693 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49693 non-null  int64  
 1   product_name   49677 non-null  object 
 2   aisle_id       49693 non-null  int64  
 3   department_id  49693 non-null  int64  
 4   prices         49693 non-null  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.9+ MB


In [7]:
# checking the shape (number of columns and rows)
df_prods.shape

(49693, 5)

### 3.0 Checking the columns with mix datatype

In [13]:
#checking fot mix data type
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)
    

product_name


### 4.0 checking for the column that contains null 

In [14]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

#### the column product_name contains 16 null values

### 5.0 Creating subset of the dataframes 

In [15]:
#Creating subset of the dataframes with null values using isnull() = True and assigning it to df_prods_nan
df_prods_nan = df_prods[df_prods['product_name'].isnull() ==True]

In [16]:
df_prods_nan.head()

product_id product_name  aisle_id  department_id  prices
33           34          NaN       121             14    12.2
68           69          NaN        26              7    11.8
115         116          NaN        93              3    10.8
261         262          NaN       110             13    12.1
525         525          NaN       109             11     1.2

In [17]:
#Checking the shape of the subset dataframes with null value
df_prods_nan.shape


(16, 5)

In [18]:
#Creating subset of the dataframes without null values using isnull() = False and assigning it to df_prods_clean
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [20]:
#Checking the shape of the subset dataframes without null value
df_prods_clean.shape

(49677, 5)

### 6.0 Checking for duplicate 

In [21]:
# checking for duplicate in the clean df_prods_clean and assigning it to df_dups
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [22]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [83]:
#Removing the duplicated value and assigning to df_prods_clean_no_dups
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()


In [84]:
df_prods_clean_no_dups


product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [254]:
# saving the df_prods_clean_no_dups.csv
df_prods_clean_no_dups.to_csv(r'Downloads/df_prods_clean_no_dups.csv')


### 7.0 Importing orders dataset

In [247]:
#Importing orders dataset into dataframe
df_ords = pd.read_csv('Downloads/CF_orders.csv')
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

##### Question : Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.Tip: Keep an eye on min and max values!

In [231]:
# checking the statistics of the ords dataframe
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

### CHECKING THE ROW AND COLUMN NUMBER

In [232]:
# Checking the shape (number of columns and rows)
df_ords.shape

(3421083, 7)

In [233]:
# Checking dataframes with mix data type for df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
    
    

 The is no mix datatype in the df_ords dataframes

### 8.0 Checking for the missing number in dataframe

question - Run a check for missing values in your df_ords dataframe. In a markdown cell, report your findings and propose an explanation for any missing values you find

In [234]:
#checking for missing value in df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

 The days_since_prior_order contains 206209 missing day, this might be due to data corruption or it might be due to the value wasnt recorded

Question - Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.

######  The missing value is in column (days_since_prior_order with 206209 missing values). the mean was calculated using the dataframe with and without Nan and the mean value for both dataframes was (11.11)  and also the median is 7.The mean is not a robust tool since it is largely influenced by outliers. The median is better suited for skewed distributions to derive at central tendency since it is much more robust and sensible 

In [235]:
# finding the mean of the import dataframe that contains the NAN
df_ords['days_since_prior_order'].mean()


11.114836226863012

In [236]:
# finding the median of the imported dataframe that contains the NAN
df_ords['days_since_prior_order'].median()

7.0

In [216]:
#Checking the mean of the dataframe withoutNAN
df_ords_without_na['days_since_prior_order'].mean()

11.114836226863012

In [194]:
# calculating the number of rows and columns
df_ords_without_na.shape

(3214874, 7)

In [237]:
# using the median for inputing 7.0 the missing NAN
df_ords['days_since_prior_order'].fillna(7.0, inplace=True)

In [238]:
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     7.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [239]:
# Creating a clean  dataframe 
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [248]:
df_ords_clean

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     7.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

 QUESTION : Run a check for duplicate values in your df_ords data.In a markdown cell, report your findings andpropose an explanation for any duplicate values you find.

In [249]:
# checking for duplicate 
df_ords_clean_no_dups = df_ords_clean.drop_duplicates()

In [250]:
df_ords_clean_no_dups

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     7.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [251]:
# checking the number of columns and rows 
df_ords_clean_no_dups.shape

(3421083, 7)

In [255]:
#saving the dataframe as df_ords_clean_no_dups
df_ords_clean_no_dups.to_csv(r'Downloads/df_ords_clean_no_dups.csv')